src/Microsoft.Azure.ServiceBus/ServiceBusConnection.cs

```diff
@@ -0,0 +1,248 @@
+            this.TransactionController = new FaultTolerantAmqpObject<Controller>(this.CreateControllerAsync, CloseController);
+        }
+
+        static void CloseConnection(AmqpConnection connection)
+        {
+            MessagingEventSource.Log.AmqpConnectionClosed(connection);
+            connection.SafeClose();
+        }
+
+        static void CloseController(Controller controller)
+        {
+            controller.Close();
+        }
+
+        async Task<AmqpConnection> CreateConnectionAsync(TimeSpan timeout)
+        {
+            var hostName = this.Endpoint.Host;
+
+            var timeoutHelper = new TimeoutHelper(timeout);
+            var amqpSettings = AmqpConnectionHelper.CreateAmqpSettings(
+                amqpVersion: AmqpVersion,
+                useSslStreamSecurity: true,
+                hasTokenProvider: true,
+                useWebSockets: TransportType == TransportType.AmqpWebSockets);
+
+            var transportSettings = CreateTransportSettings();
+            var amqpTransportInitiator = new AmqpTransportInitiator(amqpSettings, transportSettings);
+            var transport = await amqpTransportInitiator.ConnectTaskAsync(timeoutHelper.RemainingTime()).ConfigureAwait(false);
+
+            var containerId = Guid.NewGuid().ToString();
+            var amqpConnectionSettings = AmqpConnectionHelper.CreateAmqpConnectionSettings(AmqpConstants.DefaultMaxFrameSize, containerId, hostName);
+            var connection = new AmqpConnection(transport, amqpSettings, amqpConnectionSettings);
+            await connection.OpenAsync(timeoutHelper.RemainingTime()).ConfigureAwait(false);
+
+            // Always create the CBS Link + Session
+            var cbsLink = new AmqpCbsLink(connection);
+            if (connection.Extensions.Find<AmqpCbsLink>() == null)
+            {
+                connection.Extensions.Add(cbsLink);
+            }
+
+            MessagingEventSource.Log.AmqpConnectionCreated(hostName, connection);
+
+            return connection;
+        }
+
+        async Task<Controller> CreateControllerAsync(TimeSpan timeout)
+        {
+            var timeoutHelper = new TimeoutHelper(timeout);
+            var connection = await this.ConnectionManager.GetOrCreateAsync(timeoutHelper.RemainingTime());
+
+            var sessionSettings = new AmqpSessionSettings { Properties = new Fields() };
+            AmqpSession amqpSession = null;
+            Controller controller;
+
+            try
+            {
+                amqpSession = connection.CreateSession(sessionSettings);
+                await amqpSession.OpenAsync(timeoutHelper.RemainingTime());
+
+                controller = new Controller(amqpSession, timeoutHelper.RemainingTime());
+                await controller.OpenAsync(timeoutHelper.RemainingTime());
+            }
+            catch (Exception exception)
+            {
+                if (amqpSession != null)
```

Comment:
Can be reduced to `amqpSession?.CloseAsync()`

Reply:
Cannot await on 'null'. So cant reduce.